### DISCUSSION TITLE
[SPEC] Xaml Designer  For All Platforms Excluding Web.

### DISCUSSION THREAD

**Discussion by davidbuckleyni** [dotnet/maui#128] [General]
**Xaml Designer**

Android and Ios in their own projects have a Gui Editor with the conversion to Project Maui we should allow developers of both android and iOS applications to use some kind of drag and drop interface.

**# API**

Interface 
The interface should allow the end user to drag and drop buttons text and any Maui control onto the surface. The user should be able to choose from 

**StackLayout
GridLayout** 

With a Grid Design Interface should popup for us to say how many rows and columns much the same way word table gen works. Grid Lines should also appear so we can place mock data into the wire frame xaml.

![Screenshot 2020-05-29 at 20 52 40](https://user-images.githubusercontent.com/5619229/83299792-c190c480-a1ee-11ea-8bb5-cc8621eea305.png)

Once the user has crafted the user interface it should give them an option to apply constraints these should be either.

Boiler Plate code should be produced in terms of the xaml layer and also the designer file. The Boiler Plate code that is generated should give the option Either to Produce xaml Or c# for markup

**Form Wizard**
If the person already has models created their should be a way that the toooling in the editor could produce a xamrain form out of it the user should be able to specify if its 2 column or a stack layout. 

Place holder code for the form events should also be generated such as save cancel return to previous view etc add button.

This would allow the programmer to be more productive and also would allow designers to use the designer side of the xaml designer so that the designs are true representations of the photoshop file or the xd document.

Some tooling also to define zones in photo shop or xd would also be good.

**Affected**
IOS
Andriod
Tizen
Android TV
Watch Os
Apple Os
Web ( Depending on Community Stand-up Review)
## XAML Example
```xaml
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms" xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml" xmlns:d="http://xamarin.com/schemas/2014/forms/design" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" mc:Ignorable="d" x:Class="GuiEeitor.Views.NewItemPage" Title="New Item">
    <ContentPage.ToolbarItems>
        <ToolbarItem Text="Cancel" Clicked="Cancel_Clicked" />
        <ToolbarItem Text="Save" Clicked="Save_Clicked" />
    </ContentPage.ToolbarItems>
    <ContentPage.Content>
        <StackLayout Spacing="20" Padding="15">
            <Label Text="Text" FontSize="Medium" />
            <Entry Text="{Binding Item.Text}" d:Text="Item name" FontSize="Small" />
            <Label Text="Description" FontSize="Medium" />
            <Editor Text="{Binding Item.Description}" d:Text="Item description" FontSize="Small" Margin="0" />
            <Button x:Name="btnSave" Clicked="btnSave_Clicked"></Button>
        </StackLayout>
    </ContentPage.Content>
</ContentPage>


```C#
namespace GuiEeitor.Views
    {
     // Learn more about making custom code visible in the Xamarin.Forms previewer
     // by visiting https://aka.ms/xamarinforms-previewer
     [DesignTimeVisible(true)]
     public partial class NewItemPage : ContentPage
    {
        public Item Item { get; set; }

        public NewItemPage()
        {
            InitializeComponent();

            Item = new Item
            {
                Text = "Item name",
                Description = "This is an item description."
            };

            BindingContext = this;
        }

        async void Save_Clicked(object sender, EventArgs e)
        {
            MessagingCenter.Send(this, "AddItem", Item);
            await Navigation.PopModalAsync();
        }

        async void Cancel_Clicked(object sender, EventArgs e)
        {
            await Navigation.PopModalAsync();
        }

        void btnSave_Clicked(System.Object sender, System.EventArgs e)
        {
        }
     }
    }
```

**Comment by legistek** [comment]
Just curious why the title of this issue ends with "Excluding Web". Does that mean web IS expected to be a platform target for Maui? 

And if so why would it be excluded from a XAML designer?

**Comment by davidbuckleyni** [comment]
> Just curious why the title of this issue ends with "Excluding Web". Does that mean web IS expected to be a platform target for Maui?
> 
> And if so why would it be excluded from a XAML designer?

@legistek  I change the title and amended the list to include web I was just thinking of the amount of extra work at end of the day every pc has a browser a person just hits F5 and the kestrel engine is very fast for dev work that people rare complain about web being slow in dev work.

**Comment by legistek** [comment]
Hey @davidbuckleyni I think my question is a little more fundamental. Is web actually going to be a platform for Maui? That would be great news but no one has actually said it was. So that's why I asked.

**Comment by tomasfabian** [comment]
I totally agree with @legistek, it would be nice to get an answer about Web (browsers) targeting in MAUI.
What does this mean in the Affected section in the spec proposal?

> Web ( Depending on Community Stand-up Review) 

Thx

**Comment by Suriman** [comment]
I totally agree with @legistek and @tomasfabian. MAUI must support Web.

**Comment by legistek** [comment]
Clearly they're seriously contemplating web as a platform (and how could they not be?) and aren't ready to share any details. At least that's what I'm choosing to believe.

**Comment by davidbuckleyni** [comment]
Affected means which areas it should be implemented in and what we would like to see it implemented in. Of course we have had no say so if web would even be in Maui it makes since it will be.

**Comment by legistek** [comment]
@davidbuckleyni my apologies, I mistook this for an official issue from the team, following the template etc., and thought maybe you knew something we didn't. :)

**Comment by davidbuckleyni** [comment]
No I just no how to create a spec properly

**Comment by pieckenst** [comment]
Welp so far literally this discussion thread didn't have any activity for 2 years
TBH in my opinion, i personally wouldn't try even making a .net maui app unless,at some point there would be some form of visual designer available

**Comment by davidbuckleyni** [comment]
I think we should mark this close for disucssion
